go/worker/client/committee/node.go

```diff
@@ -149,22 +149,42 @@ func (n *Node) Query(ctx context.Context, round uint64, method string, args []by
 	}
 	maxMessages := dsc.Executor.MaxMessages
 
-	annBlk, err := n.commonNode.Runtime.History().GetAnnotatedBlock(ctx, round)
-	if err != nil {
-		return nil, fmt.Errorf("client: failed to fetch annotated block from history: %w", err)
-	}
+	var resolvedRound uint64
+	queryFn := func(round uint64) ([]byte, error) {
+		annBlk, err := n.commonNode.Runtime.History().GetAnnotatedBlock(ctx, round)
+		if err != nil {
+			return nil, fmt.Errorf("client: failed to fetch annotated block from history: %w", err)
+		}
+		resolvedRound = annBlk.Block.Header.Round
 
-	// Get consensus light block for state after executing block at given height.
-	lb, err := n.commonNode.Consensus.GetLightBlockForState(ctx, annBlk.Height)
-	if err != nil {
-		return nil, fmt.Errorf("client: failed to get light block at height %d: %w", annBlk.Height, err)
-	}
-	epoch, err := n.commonNode.Consensus.Beacon().GetEpoch(ctx, annBlk.Height)
-	if err != nil {
-		return nil, fmt.Errorf("client: failed to get epoch at height %d: %w", annBlk.Height, err)
+		// Get consensus light block for state after executing block at given height.
+		lb, err := n.commonNode.Consensus.GetLightBlockForState(ctx, annBlk.Height)
+		if err != nil {
+			return nil, fmt.Errorf("client: failed to get light block at height %d: %w", annBlk.Height, err)
+		}
+		epoch, err := n.commonNode.Consensus.Beacon().GetEpoch(ctx, annBlk.Height)
+		if err != nil {
+			return nil, fmt.Errorf("client: failed to get epoch at height %d: %w", annBlk.Height, err)
+		}
+
+		return hrt.Query(ctx, annBlk.Block, lb, epoch, maxMessages, method, args)
 	}
 
-	return hrt.Query(ctx, annBlk.Block, lb, epoch, maxMessages, method, args)
+	data, err := queryFn(round)
+	if protocol.IsConsensusVerifierError(err) {
```

Comment:
Could this handle a more specific error, e.g. a specific code? (I assume `VerificationFailed`)? Or are there multiple different errors that can happen in case it is failing due to finalized header not being available.

Reply:
Agreed, updated to use a more specific error, take a look.